linera-execution/src/wasm/system_api.rs

```diff
@@ -1,6 +1,574 @@
+            log::Level::Trace => tracing::trace!("{message}"),
+            log::Level::Debug => tracing::debug!("{message}"),
+            log::Level::Info => tracing::info!("{message}"),
+            log::Level::Warn => tracing::warn!("{message}"),
+            log::Level::Error => tracing::error!("{message}"),
+        }
+        Ok(())
+    }
+}
+
+/// An implementation of the system API used to access the view storage for both contracts and
+/// services.
+#[derive(Default)]
+pub struct ViewSystemApi<Caller>(PhantomData<Caller>);
+
+#[linera_witty::wit_export(package = "linera:app")]
+impl<Caller, Runtime> ViewSystemApi<Caller>
+where
+    Caller: Instance<UserData = SystemApiData<Runtime>>,
+    Runtime: BaseRuntime + Send + 'static,
+{
+    /// Creates a new promise to check if the `key` is in storage.
+    fn contains_key_new(caller: &mut Caller, key: Vec<u8>) -> Result<u32, RuntimeError> {
+        let mut data = caller.user_data_mut();
+        let promise = data
+            .runtime
+            .contains_key_new(key)
+            .map_err(|error| RuntimeError::Custom(error.into()))?;
+
+        data.register_promise(promise)
+    }
+
+    /// Waits for the promise to check if the `key` is in storage.
+    fn contains_key_wait(caller: &mut Caller, promise_id: u32) -> Result<bool, RuntimeError> {
+        let mut data = caller.user_data_mut();
+        let promise = data.take_promise(promise_id)?;
+
+        data.runtime
+            .contains_key_wait(&promise)
+            .map_err(|error| RuntimeError::Custom(error.into()))
+    }
+
+    /// Creates a new promise to read multiple entries from storage.
+    fn read_multi_values_bytes_new(
+        caller: &mut Caller,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<u32, RuntimeError> {
+        let mut data = caller.user_data_mut();
+        let promise = data
+            .runtime
+            .read_multi_values_bytes_new(keys)
+            .map_err(|error| RuntimeError::Custom(error.into()))?;
+
+        data.register_promise(promise)
+    }
+
+    /// Waits for the promise to read multiple entries from storage.
+    fn read_multi_values_bytes_wait(
+        caller: &mut Caller,
+        promise_id: u32,
+    ) -> Result<Vec<Option<Vec<u8>>>, RuntimeError> {
+        let mut data = caller.user_data_mut();
+        let promise = data.take_promise(promise_id)?;
+
+        data.runtime
+            .read_multi_values_bytes_wait(&promise)
+            .map_err(|error| RuntimeError::Custom(error.into()))
+    }
+
+    /// Creates a new promise to read a single entry from storage.
+    fn read_value_bytes_new(caller: &mut Caller, key: Vec<u8>) -> Result<u32, RuntimeError> {
+        let mut data = caller.user_data_mut();
+        let promise = data
+            .runtime
+            .read_value_bytes_new(key)
+            .map_err(|error| RuntimeError::Custom(error.into()))?;
+
+        data.register_promise(promise)
+    }
+
+    /// Waits for the promise to read a single from storage.
```

Comment:
A single… value?

Reply:
Oops! I added "entry" to match the `_new` promise constructor above. Thanks!